src/Orleans.Runtime/Messaging/MessageCenter.cs

```diff
@@ -180,25 +144,98 @@ public void SendMessage(Message msg)
             }
             else
             {
-                msg.SendingSilo ??= this.MyAddress;
-                OutboundQueue.SendMessage(msg);
+                msg.SendingSilo ??= _siloAddress;
+
+                if (stopped)
+                {
+                    log.LogInformation((int)ErrorCode.Runtime_Error_100115, "Message was queued for sending after outbound queue was stopped: {Message}", msg);
+                    SendRejection(msg, Message.RejectionTypes.Unrecoverable, "Message was queued for sending after outbound queue was stopped");
+                    return;
+                }
+
+                // Don't process messages that have already timed out
+                if (msg.IsExpired)
+                {
+                    this.messagingTrace.OnDropExpiredMessage(msg, MessagingStatisticsGroup.Phase.Send);
+                    return;
+                }
+
+                // First check to see if it's really destined for a proxied client, instead of a local grain.
+                if (TryDeliverToProxy(msg))
+                {
+                    // Message was successfully delivered to the proxy.
+                    return;
+                }
+
+                if (msg.TargetSilo is not { } targetSilo)
+                {
+                    log.LogError((int)ErrorCode.Runtime_Error_100113, "Message does not have a target silo: " + msg + " -- Call stack is: " + Utils.GetStackTrace());
+                    SendRejection(msg, Message.RejectionTypes.Unrecoverable, "Message to be sent does not have a target silo");
+                    return;
+                }
+
+                messagingTrace.OnSendMessage(msg);
+                if (targetSilo.Matches(_siloAddress))
+                {
+                    if (log.IsEnabled(LogLevel.Trace))
+                    {
+                        log.LogTrace("Message has been looped back to this silo: {Message}", msg);
+                    }
+
+                    MessagingStatisticsGroup.LocalMessagesSent.Increment();
+                    this.DispatchLocalMessage(msg);
+                }
+                else
+                {
+                    if (stopped)
+                    {
+                        log.LogInformation((int)ErrorCode.Runtime_Error_100115, "Message was queued for sending after outbound queue was stopped: {Message}", msg);
```

Comment:
You might want to file a workitem to clean up all of these calls to use LoggerMessage instead.

Reply:
Yeah, we need to do that. I've also been waiting for a logging analyzer to come out so that we can be more sure that we don't mess up the format string, but so far it hasn't materialized.